[BD-46] feat: create chip carousel

monteri opened the pull request:
## Description

Create chip carousel component

### Deploy Preview

https://deploy-preview-2378--paragon-openedx.netlify.app

## Merge Checklist

* [x] If your update includes visual changes, have they been reviewed by a designer? Send them a link to the Netlify deploy preview, if applicable.
* [x] Does your change adhere to the documented [style conventions](https://github.com/openedx/paragon/blob/master/docs/decisions/0012-css-styling-conventions)?
* [x] Do any prop types have missing descriptions in the Props API tables in the documentation site (check deploy preview)?
* [x] Were your changes tested using all available themes (see theme switcher in the header of the deploy preview, under the "Settings" icon)?
* [x] Were your changes tested in the `example` app?
* [x] Is there adequate test coverage for your changes?
* [x] Consider whether this change needs to reviewed/QA'ed for accessibility (a11y). If so, please add `wittjeff` and `adamstankiewicz` as reviewers on this PR.

## Post-merge Checklist

* [x] Verify your changes were released to [NPM](https://www.npmjs.com/package/@edx/paragon) at the expected version.
* [x] If you'd like, [share](https://github.com/openedx/paragon/discussions/new?category=show-and-tell) your contribution in [#show-and-tell](https://github.com/openedx/paragon/discussions/categories/show-and-tell).
* [x] 🎉 🙌 Celebrate! Thanks for your contribution.

openedx-webhooks commented:
Thanks for the pull request, @monteri!


When this pull request is ready, tag your edX technical lead.


<!-- comment:welcome-blended -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

PKulkoRaccoonGang commented on src/ChipCarousel/ChipCarousel.scss:
```diff
@@ -0,0 +1,24 @@
+  }
+}
+
+.pgn__chip-carousel__right-control,
```
These selectors are children `.pgn__chip-carousel`?

monteri replied:
Yes, but if we don't need nesting, it is appropriate to avoid it to have more readable styles.

adamstankiewicz replied:
Yeah, I'd say the nesting isn't a hard requirement given we have the convention of BEM class names (mostly) prefixed with `pgn__{componentname}`. Unlikely for collisions to happen, though still possible.

PKulkoRaccoonGang commented on src/ChipCarousel/ChipCarousel.test.jsx:
```diff
@@ -0,0 +1,68 @@
+import React from 'react';
+import { render, fireEvent } from '@testing-library/react';
+import { IntlProvider } from 'react-intl';
+import ChipCarousel from './index';
```
```suggestion
import ChipCarousel from '.';
```

PKulkoRaccoonGang commented on src/ChipCarousel/README.md:
```diff
@@ -0,0 +1,92 @@
+notes: |
+---
+
+``ChipCarousel`` component creates a scrollable horizontal block of chips with buttons for navigating to the previous and next set of chips.
```
```suggestion
The ``ChipCarousel`` component creates a scrollable horizontal block of chips with buttons for navigating to the previous and next set of chips.
```

PKulkoRaccoonGang commented on src/ChipCarousel/README.md:
````diff
@@ -0,0 +1,92 @@
+  />
+```
+
+## `Chip` Carousel
````
Interactive Сhip Сarousel maybe

monteri replied:
I just forgot to delete this example

PKulkoRaccoonGang commented on src/ChipCarousel/README.md:
```diff
@@ -0,0 +1,92 @@
+        </div>
+        <div ref={setOverflowRef} className="d-flex">
+          <OverflowScroll.Items>
+            <Chip iconAfter={Close}>New</Chip>
```
I propose to simulate an array in this way. This will help us keep the example compact.

```
{[...Array(12).keys()].map(item => (
  <Chip key={item + 1} iconAfter={Close}>New</Chip>
))}
```

PKulkoRaccoonGang commented on src/ChipCarousel/index.tsx:
```diff
@@ -0,0 +1,140 @@
+import { useIntl } from 'react-intl';
+import classNames from 'classnames';
+// @ts-ignore
+import { OverflowScroll, OverflowScrollContext } from '../OverflowScroll';
```
Let's try to ignore the necessary import rule for the whole file, this will help to avoid numerous `@ts-ignore`

monteri replied:
I would do it as a separate PR

adamstankiewicz replied:
[clarification] Once the Paragon types PR merges in, I'm assuming these `@ts-ignore` comment would become unnecessary?

PKulkoRaccoonGang commented on src/ChipCarousel/messages.js:
```diff
@@ -0,0 +1,16 @@
+
+const messages = defineMessages({
+  scrollToPrevious: {
+    id: 'pgn.ChipCarousel.scrollToPrevious',
```
We do not use camelСase in the declaration of transfer identifiers, let's use exclusively lowercase identifiers

monteri replied:
I just stick to the other `id` in the Paragon, do you think we need to change to lowercase all of them?

PKulkoRaccoonGang replied:
`id: 'pgn.chip-carousel.scrollToPrevious',` maybe

adamstankiewicz replied:
> We do not use camelСase in the declaration of transfer identifiers, let's use exclusively lowercase identifiers

@PKulkoRaccoonGang [clarification] Can you expand on this comment a bit more?

adamstankiewicz replied:
@PKulkoRaccoonGang @monteri After discussing, we decided to stick with the camelCase message ID format (it's used through frontend-app-learning and some other MFEs. Up to the owning team of repo to decide the standard).

PKulkoRaccoonGang commented on src/ChipCarousel/messages.js:
```diff
@@ -0,0 +1,16 @@
+    defaultMessage: 'Scroll to previous',
+    description: 'Accessibility text describing the action for navigating a scrollable carousel to the previous element',
+  },
+  scrollToNext: {
```
Also in other translations, I noticed that we use a slightly different format for recording the key for translations. I suggest sticking to uniformity.

`  'subscription.details.program.type.microMasters': {
    id: 'subscription.details.program.type.microMasters',
    defaultMessage: 'MicroMasters',
    description: 'Label to render the MicroMasters Certificate program type.',
  },`

PKulkoRaccoonGang commented on src/index.scss:
```diff
@@ -11,6 +11,7 @@
 @import "./Collapsible/Collapsible";
 @import "./CloseButton/CloseButton";
 @import "./Chip/Chip";
+@import "./ChipCarousel/ChipCarousel";
```
Curious moment. It seems to me that if the style files are called index, we can remove the duplication of the name from the path here

For example: `@import "./ChipCarousel"`

monteri replied:
Yeah, I think it can be considered as separate issue

PKulkoRaccoonGang commented on src/ChipCarousel/_variables.scss:
```diff
@@ -0,0 +1 @@
+$chip-carousel-controls-top-offset:   -3px !default;
```
It will be necessary to add a new token :)

adamstankiewicz commented on src/ChipCarousel/README.md:
```diff
@@ -0,0 +1,73 @@
+            setValue: setOffset,
+            range: {
+              min: 0,
+              max: offsetType === 'percentage' ? 100 : 1000,
+              step: offsetType === 'percentage' ? 1 : 50,
```
On page load, `offsetType` is defaulted to "fixed" at a value of 120. However, if the user switches to "percentage", 120 > the `max` of 100, so the percentage offset changes to 100%, which doesn't quite demonstrate the percentage increment option effectively.

I wonder if we should reset the `offset` value when switching between `offsetType`s so the offset values make sense for each type upon switching?

monteri replied:
Added function to be more user-friendly when `offsetType` is switched. It calculates ration and sets relatively equal value between `fixed` and `percentage`.

adamstankiewicz commented on src/ChipCarousel/README.md:
```diff
@@ -0,0 +1,73 @@
+            value: gap,
+            setValue: setGap,
+            range: { min: 0, max: 6, step: 0.5 },
+            name: 'offset'
```
Should this be named `gap` instead?

monteri replied:
updated to `gap`

adamstankiewicz commented on src/ChipCarousel/README.md:
```diff
@@ -0,0 +1,73 @@
+            value: gap,
+            setValue: setGap,
+            range: { min: 0, max: 6, step: 0.5 },
+            name: 'offset'
```
[curious] It looks like there's no difference in the gap between `0` and `0.5`; is this expected?

<img width="942" alt="image" src="https://github.com/openedx/paragon/assets/2828721/ba140860-abd7-46b6-96c4-230970142d5e">

<img width="940" alt="image" src="https://github.com/openedx/paragon/assets/2828721/803199ab-03e6-4975-8a8e-2dcf1dc555c7">

monteri replied:
Yes, probably I have talked about this case that there is no `.5` `spacer` value.
```
$spacers: map-merge(
  (
    0: 0,
    1: ($spacer * .25),
    1\.5: ($spacer * .375),
    2: ($spacer * .5),
    2\.5: ($spacer * .75),
    3: $spacer,
    3\.5: ($spacer * 1.25),
    4: ($spacer * 1.5),
    4\.5: ($spacer * 2),
    5: ($spacer * 3),
    5\.5: ($spacer * 4),
    6: ($spacer * 5),
  ),
  $spacers
);
```

adamstankiewicz replied:
Ah, yes. Thank you for the reminder. Noted :)

adamstankiewicz commented on src/OverflowScroll/OverflowScroll.jsx:
```diff
@@ -84,7 +84,9 @@ OverflowScroll.propTypes = {
   /** Callback function for when the user scrolls to the next element. */
   onScrollNext: PropTypes.func,
-  offset: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+  /** A value specifying the distance the scroll should move. */
```
🎉

adamstankiewicz commented on src/ChipCarousel/ChipCarousel.test.jsx:
```diff
@@ -0,0 +1,68 @@
+  it('should render the carousel correctly', () => {
+    render(<TestingChipCarousel />);
+
+    const carousel = document.getElementsByClassName('pgn__chip-carousel');
```
Is there a way around needing to rely on the implementation details of the component in these tests by relying more on the `@testing-library/react` (RTL) API?

This is kind of using RTL almost as if it were Enzyme by relying on implementation details like class names. Perhaps we rely more on `data-testid` attributes here for selections?

```jsx
function TestingChipCarousel(props) {
  return (
    <IntlProvider locale="en">
      <ChipCarousel data-testid="chip-carousel" ariaLabel={ariaLabel} items={items} {...props} />
    </IntlProvider>
  );
}

render(<TestingChipCarousel />);

const carousel = screen.getByTestId('chip-carousel');
```

Similarly for the `chipItems` below, we could do something like add a `data-testid="chip-component"` on each child object passed to `ChipCarousel` such that we could query using RTL like: `screen.queryByTestId('chip-component')`

Reference: https://kentcdodds.com/blog/testing-implementation-details

monteri replied:
Agree on this point. Refactored tests to use `data-testid`

adamstankiewicz commented on src/ChipCarousel/ChipCarousel.test.jsx:
```diff
@@ -0,0 +1,68 @@
+
+    const chipItems = document.getElementsByClassName('pgn__chip');
+    for (let i = 0; i < chipItems.length; i++) {
+      fireEvent.click(chipItems[i]);
```
Typically, `userEvent.click` is preferred over using `fireEvent` directly. See https://stackoverflow.com/a/71031849.

monteri replied:
updated to use `userEvent.click`

adamstankiewicz reviewed (approved):
LGTM! Just looks like it might need a rebase with `master`, though.

monteri commented:
> Yeah, I'd say the nesting isn't a hard requirement given we have the convention of BEM class names (mostly) prefixed with pgn__{componentname}. Unlikely for collisions to happen, though still possible.

Updated to use nesting in this case. Also noticed that when we changed `componentName.scss` to `index.scss` we forgot to update `component-generator`. Made quick fix as part of this PR

adamstankiewicz commented:
> Updated to use nesting in this case. Also noticed that when we changed `componentName.scss` to `index.scss` we forgot to update `component-generator`. Made a quick fix as part of this PR

Good catch!

openedx-webhooks commented:
@monteri 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/paragon&entry.867055334=https://github.com/openedx/paragon/pull/2378&entry.1484655318=https://github.com/monteri&entry.752974735=2023-06-16+11:47&entry.1917517419=2023-08-04+12:16&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

edx-semantic-release commented:
:tada: This PR is included in version 20.46.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@edx/paragon/v/20.46.0)
- [GitHub release](https://github.com/openedx/paragon/releases/tag/v20.46.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
